lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,307 @@
+    .getCanonicalElementFromIdentifier(assignment.leftHandSide);
+
+bool _isLoopStatement(AstNode node) =>
+    node is DoStatement ||
+    node is ForEachStatement ||
+    node is ForStatement ||
+    node is WhileStatement;
+
+List<Expression> _splitConjunctions(Expression expression) {
+  if (expression is BinaryExpression &&
+      expression.operator.type == TokenType.AMPERSAND_AMPERSAND) {
+    return _splitConjunctions(expression.leftOperand)
+      ..addAll(_splitConjunctions(expression.rightOperand));
+  } else if (expression is ParenthesizedExpression) {
+    return _splitConjunctions(expression.expression);
+  }
+  return [expression];
+}
+
+class ConditionScope {
+  final environment = <_ExpressionBox>[];
+  final ConditionScope outer;
+
+  ConditionScope(this.outer);
+
+  void add(_ExpressionBox e) {
+    if (e != null) {
+      environment.add(e);
+    }
+  }
+
+  void addAll(Iterable<_ExpressionBox> expressions) {
+    environment.addAll(expressions);
+  }
+
+  Iterable<Expression> getExpressions(Iterable<Element> elements, bool value) {
+    final expressions = <Expression>[];
+    _recursiveGetExpressions(this, expressions, elements, value);
+    return expressions;
+  }
+
+  Iterable<_ExpressionBox> getUndefinedExpressions() =>
+      environment.where((e) => e is _UndefinedExpression);
+
+  static void _recursiveGetExpressions(ConditionScope conditionScope,
+      List<Expression> expressions, Iterable<Element> elements, bool value) {
+    if (conditionScope == null) {
+      return;
+    }
+    for (final element in conditionScope.environment.reversed) {
+      if (element.haveToStop(elements)) {
+        return;
+      }
+      if (element is _ConditionExpression && element.value == value) {
+        expressions.add(element.expression);
+      }
+    }
+    _recursiveGetExpressions(
+        conditionScope.outer, expressions, elements, value);
+  }
+}
+
+abstract class ConditionScopeVisitor extends UnifyingAstVisitor {
+  ConditionScope outerScope;
+
+  Iterable<Expression> getFalseExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements, value: false);
+
+  Iterable<Expression> getTrueExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _addElementToEnvironment(new _UndefinedExpression(_getLeftElement(node)));
+    node.visitChildren(this);
+  }
+
+  @override
+  visitBlockFunctionBody(BlockFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  void visitCondition(Expression node);
+
+  @override
+  visitDoStatement(DoStatement node) {
+    _addScope();
+    visitCondition(node.condition);
+    node.visitChildren(this);
+    _propagateUndefinedExpressions(_removeLastScope());
+  }
+
+  @override
+  visitEmptyFunctionBody(EmptyFunctionBody node) {
```

Comment:
Given that an empty function body doesn't have any children, this method should be unnecessary.